packages/analytics-engine/src/engine.ts

```diff
@@ -0,0 +1,289 @@
+import { TextDecoder } from "util";
+import type { SqliteDB } from "@miniflare/shared";
+import analytics from "./analytics";
+import buildSQLFunctions, { isDate } from "./functions";
+
+export type Format = "JSON" | "JSONEachRow" | "TabSeparated";
+
+export type MetaType = "DateTime" | "String" | "Float64";
+
+export interface AnalyticsEngineEvent {
+  readonly doubles?: number[]; // up to 20
+  readonly blobs?: (ArrayBuffer | string | null)[]; // up to 20, max sum of all blobs: 5kb
+  readonly indexes?: string[]; // 0 or 1
+}
+
+interface DataPoint {
+  dataset: string;
+  index1?: string;
+  double1?: number;
+  double2?: number;
+  double3?: number;
+  double4?: number;
+  double5?: number;
+  double6?: number;
+  double7?: number;
+  double8?: number;
+  double9?: number;
+  double10?: number;
+  double11?: number;
+  double12?: number;
+  double13?: number;
+  double14?: number;
+  double15?: number;
+  double16?: number;
+  double17?: number;
+  double18?: number;
+  double19?: number;
+  double20?: number;
+  blob1?: string | null;
+  blob2?: string | null;
+  blob3?: string | null;
+  blob4?: string | null;
+  blob5?: string | null;
+  blob6?: string | null;
+  blob7?: string | null;
+  blob8?: string | null;
+  blob9?: string | null;
+  blob10?: string | null;
+  blob11?: string | null;
+  blob12?: string | null;
+  blob13?: string | null;
+  blob14?: string | null;
+  blob15?: string | null;
+  blob16?: string | null;
+  blob17?: string | null;
+  blob18?: string | null;
+  blob19?: string | null;
+  blob20?: string | null;
+}
+
+export interface ResponseData {
+  [key: string]: number | string;
+}
+
+export interface FormatJSON {
+  meta: { [key: string]: MetaType };
+  data: ResponseData[];
+  rows: number;
+}
+
+const decoder = new TextDecoder();
+
+export class AnalyticsEngine {
+  readonly #dataset: string;
+  readonly #db: SqliteDB;
+
+  constructor(dataset: string, db: SqliteDB) {
+    this.#dataset = dataset;
+    this.#db = db;
+    db.exec(analytics.replaceAll("{{BINDING}}", dataset));
+    buildSQLFunctions(db);
+  }
+
+  writeDataPoint({
+    indexes = [],
+    doubles = [],
+    blobs = [],
+  }: AnalyticsEngineEvent): void {
+    for (const blob of blobs) {
+      if (
+        blob !== null &&
+        typeof blob !== "string" &&
+        !(blob instanceof ArrayBuffer)
+      ) {
+        throw new Error('"blobs" may only be an ArrayBuffer, string, or null.');
+      }
+    }
+    for (const double of doubles) {
+      if (typeof double !== "number") {
+        throw new Error('"doubles" may only contain numbers.');
+      }
+    }
+    // convert arrayBuffers if they exist
+    const _blobs = blobs.map((blob) => {
+      if (blob instanceof ArrayBuffer) {
+        return decoder.decode(new Uint8Array(blob));
+      }
+      return blob;
```

Comment:
I can switch to `BLOB`s no problem. The change is fine, but:
* If I store as ArrayBuffer it will fail (can only be string, null, or Uint8Array with better-sqlite)
* If I store as Uint8Array, the return is a Buffer when doing searches when they should be strings
* https://developers.cloudflare.com/analytics/analytics-engine/sql-api/#table-structure -> Table is actually text which is why I assumed it should be treated as such.